src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/Frame.cs

```diff
@@ -552,9 +572,22 @@ public void Flush()
             Output.Flush();
         }
 
-        public async Task FlushAsync(CancellationToken cancellationToken)
+        public Task FlushAsync(CancellationToken cancellationToken)
         {
-            await InitializeResponse(0);
+            var initializeTask = InitializeResponseMayAwait(0);
+            // If return is TaskCache.CompletedTask no awaiting is required
+            if (!ReferenceEquals(initializeTask, TaskCache.CompletedTask))
+            {
+                return FlushAsyncAwaited(initializeTask, cancellationToken);
+            }
+
+            return Output.FlushAsync(cancellationToken);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
The `MoveNext()` method is 275 bytes of il so wouldn't be inlined; but the init `FlushAsyncAwaited` call is 73 bytes of il so it is possible... (though unlikely)